Stable error handling for loaded Cordova plugins

mathiasscheffe opened the pull request:
**Platforms affected**
All

**Motivation and Context**
When developing a large Cordova app in a distributed team occasionally a developer commits a Cordova plugin which contains an issue which directly results in a javaScript exception on plugin load. 
Before the fix the Cordova startup crashes and the app results in a whitescreen.

**Description**
The fix makes handlePluginsObject more stable. The function keeps track of the loaded modules. If a load error occurs for a module the module is removed from the modules list. Additionally an alert is displayed. This makes it way easier for developers to track and fix the error.

<!--
Please make sure the checklist boxes are all checked before submitting the PR. The checklist is intended as a quick reference, for complete details please see our Contributor Guidelines:

http://cordova.apache.org/contribute/contribute_guidelines.html

Thanks!
-->

### Platforms affected



### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->



### Description
<!-- Describe your changes in detail -->



### Testing
<!-- Please describe in detail how you tested your changes. -->



### Checklist

- [ ] I've run the tests to see all new and existing tests pass
- [ ] I added automated test coverage as appropriate for this change
- [ ] Commit is prefixed with `(platform)` if this change only applies to one platform (e.g. `(android)`)
- [ ] If this Pull Request resolves an issue, I linked to the issue in the text above (and used the correct [keyword to close issues using keywords](https://help.github.com/articles/closing-issues-using-keywords/))
- [ ] I've updated the documentation if necessary

raphinesse commented on src/common/pluginloader.js:
```diff
@@ -79,19 +79,63 @@ function onScriptLoadingComplete (moduleList, finishPluginLoading) {
     for (var i = 0; i < moduleList.length; i++) {
-        injectIfNecessary(moduleList[i].id, path + moduleList[i].file, scriptLoadedCallback);
+        var moduleId = moduleList[i].id;
+        var me = this;
+        // bound function to have the module id when the error occurs.
+        var boundErrorCallback = scriptLoadedErrorCallback.bind(me, moduleId);
```
```suggestion
        // bound function to have the module id when the error occurs.
        var boundErrorCallback = scriptLoadedErrorCallback.bind(null, moduleId);
```
Since `scriptLoadedErrorCallback` does not use `this`, there's no need to bind it.

mathiasscheffe replied:
This looks fine for me as well.
My intention was to create as little surprises as possible for future developers. Therefore I wanted to keep the this pointer intact. But it is not needed now and will also work in the way you suggested.

raphinesse commented on src/common/pluginloader.js:
```diff
@@ -79,19 +79,63 @@ function onScriptLoadingComplete (moduleList, finishPluginLoading) {
     }
+
+    function callOnScriptLoadingComplete () {
+        modulesWithProblems.forEach(function (elem) {
+            var foundIndex = -1;
+            var modulesLength = moduleList.length;
+            var i = 0;
+            // find module with load problems in module list. Use while loop to exit iteration early.
+            while (i < modulesLength && foundIndex === -1) {
+                if (moduleList[i].id === elem) {
+                    foundIndex = i;
+                }
+                i++;
+            }
+            if (foundIndex !== -1) {
+                // delete module with load problem from module list
+                moduleList.splice(foundIndex, 1);
+            }
+        });
+        onScriptLoadingComplete(moduleList, finishPluginLoading);
```
```suggestion
        var loadedModules = moduleList.filter(function (m) {
            return modulesWithProblems.indexOf(m.id) === -1;
        });
        onScriptLoadingComplete(loadedModules, finishPluginLoading);
```
That should get the job done as well while making the intention clearer. Plus we do not modify the original array.

mathiasscheffe replied:
I can also agree to this version. Reasons why I decided to implement differently is that
the version I implemented is faster as it in the outer loop only iterates over the smaller array of modules with problems. If there are no problems then no code is executed. But this does not matter that much as the module list is most likely just containing <100 elements. 
I'll wait on your comment to that and I am open to accept the suggestion.

raphinesse replied:
I don't expect any noticeable slowdown due to the single iteration over all modules. If that should turn out to be a problem, we can still fast-track the no-error case by doing
```js
var loadedModules = modulesWithProblems.length === 0
    ? moduleList
    : moduleList.filter(function (m) {
        return modulesWithProblems.indexOf(m.id) === -1;
    });
```

mathiasscheffe commented:
Regarding one consolidated error message I had two thoughts in mind:
1. small form factors: When you are running cordova on a phone and then get e.g. 10 modules with an error the message will most likely be larger than the screen size. This will then bring scrolling and awkward layout.
2. probability: In the scenarios I know from the past year a crash while loading a plugin mostly occurs when you develop plugins yourself and in your last builds a mistake was checked in. It is very unlikely that a huge list of plugins fail loading.

Before the PR is not accepted I will create a consolidated error message but I feel more comfortable with the existing solution.

mathiasscheffe commented:
>  Unit tests for this feature would be nice as well.

Do you have any suggestions on how to hook into the functionality? This will save me some debugging. The existing unit tests are of higher level.
I can make time in 2 weeks to implement unit tests.

raphinesse commented:
@mathiasscheffe Good points about the error messages. I think having separate messages should be fine then.

> Do you have any suggestions on how to hook into the functionality? This will save me some debugging. The existing unit tests are of higher level.

I think you should be able to define a fake `cordova/plugin_list` module via `cordova.define` (see `test.require.js` for examples) with some faulty plugin modules, spy on `window.alert` and `console.error` and check that they get called when loading plugin modules.

> I can make time in 2 weeks to implement unit tests.

No hurry. All in due time.

raphinesse commented:
Another important point that should be considered is the following: I see the appeal of alert during development, so you notice your mistake ASAP. But under no circumstances should Cordova show an alert in a released App. Network or file system issues could cause this in the current version of this PR AFAICT. What can we do to prevent this?

mathiasscheffe commented:
> Another important point that should be considered is the following: I see the appeal of alert during development, so you notice your mistake ASAP. But under no circumstances should Cordova show an alert in a released App. Network or file system issues could cause this in the current version of this PR AFAICT. What can we do to prevent this?

I have changed the approach here. I have removed the alert() and have left a comment.
In the Cordova after_platform_add hook each app can use this comment as a search & replace marker and inline an app specific error handling.
I will change our internal app DEV build to inline an alert. For the appstore release build I will do something which does more suit the end user.
Hope this is ok.


P.S.: Will make time until end of week to also implement unit tests.

mathiasscheffe commented:
The pull request now also contains unit tests for the newly added functionality.
If you want to debug the unit tests you need a modified karma.conf.js and the build tools need to be adapted to skip creation of the code instrumentation. I have appended these patched files for development purposes as tar.gz.
[developmentFiles.tar.gz](https://github.com/apache/cordova-js/files/4522196/developmentFiles.tar.gz)
